feat: Initial version of Java getting started page (#184)
open-feature/docs.openfeature.dev: docs/getting-started/java.md

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

# Get Started with the OpenFeature Java SDK and Spring Boot

## Introduction

This walk-through teaches you the basics of using OpenFeature in Java on top of Spring Boot Web application.

You'll learn how to:

- Integrate OpenFeature Java SDK
- Install and configure an OpenFeature provider
- Perform basic feature flagging

## Requirements

This walk-through assumes that:

- You have a basic knowledge of Java and Spring Boot
- You have Java 8 or later
- You have Docker installed and running on the host system

## Walk-through

### Step 1: Create a Spring Boot Web application

To get started, visit [Spring Initializer](https://start.spring.io/). This website allows you to download a pre-configured
Spring boot application. 

> For our use case, add **Spring Web** from dependencies section and set project metadata to your desired values.

Reviewer: I would recommend defining exactly what settings should be used.

> Download the application using **GENERATE** button. Extract the archive and open it in your favourite editor.

Author: I will stick to US :)

### Step 2: Add dependencies

Based on the selected build system for Spring Boot application, add OpenFeature Java SDK and flagd provider dependencies
to existing dependency management configuration. 

> Below examples show dependencies for Maven and Gradle and contains the latest 

Author: No this is not possible and not recommended. We need to update versions or cross-link compatible versions where possible.

versions at the time of writing this walk-through.

<Tabs groupId="dependency">
<TabItem value="maven" label="Apache Maven">

```xml
<dependency>
    <groupId>dev.openfeature</groupId>
    <artifactId>sdk</artifactId>
    <version>1.0.0</version>
</dependency>
<dependency>
    <groupId>dev.openfeature.contrib.providers</groupId>
    <artifactId>flagd</artifactId>
    <version>0.5.1</version>
</dependency>
```

</TabItem>
<TabItem value="groovy" label="Gradle Groovy DSL">

```groovy
implementation 'dev.openfeature:sdk:1.0.0'
implementation 'dev.openfeature.contrib.providers:flagd:0.5.1'
```
</TabItem>
</Tabs>

### Step 3: Add code

> Our simple web server contains a single endpoint `/hello`. Let's add required codes to get this up and running with

Reviewer: This section was a bit confusing for me at first. It made me think that the generator above created this endpoint.

> Our simple web server contains a single endpoint `/hello`. Let's add required codes to get this up and running with

Author: I will restructure content to solve this

recommended best practices. 

Please note that you may have your own package names and package naming in sample codes will require alterations.

> `OpenFeatureAPI` is the main access point to OpenFeature SDK. This object is designed to act as a singleton. Hence, let's 

Reviewer: > Hence, let's define a simple bean configuration class `OpenFeatureBeans`. This definition allows us to dependency inject `OpenFeatureAPI` singleton to desired Spring components.

How about something like this?

Let's define a simple bean configuration class called `OpenFeatureBeans` that will allows us to inject the `OpenFeatureAPI` into desired Spring components.

> `OpenFeatureAPI` is the main access point to OpenFeature SDK. This object is designed to act as a singleton. Hence, let's 

Author: Yes, this short and precise explanation. I will add this change

define a simple bean configuration class `OpenFeatureBeans`. This definition allows us to dependency inject `OpenFeatureAPI` 
singleton to desired Spring components.

```java

> package demo;

Author: Let's add realistic package names matching Spring initializer ex:- com.demo

        openFeatureAPI.setProvider(new FlagdProvider());

        return openFeatureAPI;
    }
}
```

Next, we can add our REST endpoint definition with dependency injection for `OpenFeatureAPI` instance,

```java
package demo;
import dev.openfeature.sdk.Client;
import dev.openfeature.sdk.OpenFeatureAPI;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RestController;

@RestController
public class RestHello {
    private final OpenFeatureAPI openFeatureAPI;

    @Autowired
    public RestHello(OpenFeatureAPI OFApi) {
        this.openFeatureAPI = OFApi;
    }
    
    @GetMapping("/hello")
    public String getHello() {
        final Client client = openFeatureAPI.getClient();

        // Evaluate welcome-message feature flag
        if (client.getBooleanValue("welcome-message", false)) {
            return "Hello, welcome to website.!";
        }

        return "Hello.!";
    }
}
```

At this point, we are ready to run the initial version of our application.

### Step 4: Run the initial application

Let's compile and run the application.

```shell

> ./gradlew clean build

Author: Add maven option as we already have maven dependencies above

java -jar build/libs/javaDemo-0.0.1-SNAPSHOT.jar
```

> Note that you may have a different `.jar` name based on your build configurations.

Now you can visit the url [http://localhost:8080/hello](http://localhost:8080/hello) and observe the message **Hello.!**.

> At this point, we are falling back to default response as flagd SDK cannot connect to any feature flag provider. Next, let's 

Author: Change wording giving the meaning - flagd was selected as the provider and it is already there in the code. But we did not configure the provider, hence by default we fall back to default value

Author: Updated in the latest commit

Download the application using **GENERATE** button. Extract the archive and open it in your favourite editor.

### Step 2: Add dependencies

Based on the selected build system for Spring Boot application, add OpenFeature Java SDK and flagd provider dependencies
to existing dependency management configuration. 

Below examples show dependencies for Maven and Gradle and contains the latest 
versions at the time of writing this walk-through.

<Tabs groupId="dependency">
<TabItem value="maven" label="Maven">

```xml
<dependency>
    <groupId>dev.openfeature</groupId>
    <artifactId>sdk</artifactId>
    <version>1.0.0</version>
</dependency>
<dependency>
    <groupId>dev.openfeature.contrib.providers</groupId>
    <artifactId>flagd</artifactId>
    <version>0.5.1</version>
</dependency>
```

</TabItem>
<TabItem value="gradle" label="Gradle">

```groovy
implementation 'dev.openfeature:sdk:1.0.0'
implementation 'dev.openfeature.contrib.providers:flagd:0.5.1'
```
</TabItem>
</Tabs>

### Step 3: Add code

Our simple web server contains a single endpoint `/hello`. Let's add required codes to get this up and running with
recommended best practices. 

> Please note that you may have your own package names and package naming in sample codes will require alterations.

Author: Yes correct, this will be removed

`OpenFeatureAPI` is the main access point to OpenFeature SDK. This object is designed to act as a singleton. Hence, let's 
define a simple bean configuration class `OpenFeatureBeans`. This definition allows us to dependency inject `OpenFeatureAPI` 
singleton to desired Spring components.

> ```java

Reviewer: Could you explicitly call out what the file should be named and where it should live?

> package com.demo;

Reviewer: Update package name based on the recommendations above.

@Configuration
public class OpenFeatureBeans {

    @Bean
    public OpenFeatureAPI OpenFeatureAPI() {
        final OpenFeatureAPI openFeatureAPI = OpenFeatureAPI.getInstance();
        
        // Use flagd as the OpenFeature provider and use default configurations
        openFeatureAPI.setProvider(new FlagdProvider());

        return openFeatureAPI;
    }
}
```

Next, we can add our REST endpoint definition with dependency injection for `OpenFeatureAPI` instance,

```java
package com.demo;

import dev.openfeature.sdk.Client;
import dev.openfeature.sdk.OpenFeatureAPI;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RestController;

@RestController
public class RestHello {
    private final OpenFeatureAPI openFeatureAPI;

    @Autowired
    public RestHello(OpenFeatureAPI OFApi) {
        this.openFeatureAPI = OFApi;
    }
    
    @GetMapping("/hello")
    public String getHello() {
        final Client client = openFeatureAPI.getClient();

        // Evaluate welcome-message feature flag
        if (client.getBooleanValue("welcome-message", false)) {

>             return "Hello, welcome to website.!";

Reviewer: ```suggestion
            return "Hello, welcome to website!";
```

        }

>         return "Hello.!";

Reviewer: ```suggestion
        return "Hello!";
```

    }
}
```

At this point, we are ready to run the initial version of our application.

### Step 4: Run the initial application

Let's compile and run the application.

<Tabs groupId="running">
<TabItem value="maven" label="Maven">

```shell
mvn clean install
java -jar target/demo-0.0.1-SNAPSHOT.jar
```
</TabItem>
<TabItem value="gradle" label="Gradle">

```shell
./gradlew clean build
java -jar build/libs/demo-0.0.1-SNAPSHOT.jar
```
</TabItem>
</Tabs>

> Note that you may have a different `.jar` name based on your build configurations.

> Now you can visit the url [http://localhost:8080/hello](http://localhost:8080/hello) and observe the message **Hello.!**.

Reviewer: ```suggestion
Now you can visit the url [http://localhost:8080/hello](http://localhost:8080/hello) and observe the message **Hello!**.
```

At this point, we are falling back to the default response. By default, OpenFeature will return the default value when
there are no providers configured or running. In our code, we used flagd provider, yet we did not configure and run it. 

### Step 5: Configure and run FlagD

Create a new file named `flags.json` and add the following JSON. Notice that there's a flag called `welcome-message` which matches the flag
key defined in [step 3](#step-3-add-code). The `welcome-message` flag has `on` and `off` variants that return `true` and `false` respectively.
The state property controls whether the feature flag is active or not. Finally, the defaultVariant property controls the variant that should be returned. 
In this case, the defaultVariant is `off`, therefore the value `false` would be returned.

```json
{
  "$schema": "https://raw.githubusercontent.com/open-feature/schemas/main/json/flagd-definitions.json",
  "flags": {
    "welcome-message": {
      "variants": {
        "on": true,
        "off": false
      },
      "state": "ENABLED",
      "defaultVariant": "off"
    }
  }
}
```

> > NOTE: This configuration is specific for flagd and vary across other providers.

Reviewer: ```suggestion
> NOTE: This configuration is specific for flagd and varies across providers.
```

I'm not sure what the best phrasing is here. @toddbaert any recommendations?

Next, pull the latest flagd image,

```shell

> docker pull ghcr.io/open-feature/flagd:latest

Reviewer: I don't think this step in necessary. The Docker run command should fetch the latest image.

Author: Yes correct, will remove this part.

```sh
docker run -p 8013:8013 -v $(pwd)/:/etc/flagd/ -it ghcr.io/open-feature/flagd:latest start --uri /etc/flagd/flags.json
```

### Step 7: Rerun application

Let's rerun our Java Sprint Boot application,

<Tabs groupId="running">
<TabItem value="maven" label="Maven">

```shell
mvn clean install
java -jar target/demo-0.0.1-SNAPSHOT.jar
```
</TabItem>
<TabItem value="gradle" label="Gradle">

```shell
./gradlew clean build
java -jar build/libs/demo-0.0.1-SNAPSHOT.jar
```
</TabItem>
</Tabs>

> Note that you may have a different `.jar` name based on your build configurations.

Revisit the endpoint [http://localhost:8080/hello](http://localhost:8080/hello) and you will see the same value.

Now let's change the feature flag in our `flags.json`, making `defaultVariant` to `on`

```diff
{
  "$schema": "https://raw.githubusercontent.com/open-feature/schemas/main/json/flagd-definitions.json",
  "flags": {
    "welcome-message": {
      "variants": {
        "on": true,
        "off": false
      },
      "state": "ENABLED",
-      "defaultVariant": "off"
+      "defaultVariant": "on"
    }
  }
}
```

> Revisit the endpoint [http://localhost:8080/hello](http://localhost:8080/hello) and you will be greeted with `Hello, welcome to website.!`

Reviewer: ```suggestion
Revisit the endpoint [http://localhost:8080/hello](http://localhost:8080/hello) and you will be greeted with `Hello, welcome to website!`
```

## Conclusion

This walk-through introduced you to the OpenFeature Java SDK and how it can be easily integrate into well known frameworks
such as Spring Boot. It covered how a provider can be configured to perform the flag evaluation and introduced basic feature 

> flagging concepts. It also showcased how feature flags can be updated at runtime, without requiring a coding change and a redeployment.

Reviewer: ```suggestion
flagging concepts. It also showcased how feature flags can be updated at runtime, without requiring a code change and a redeployment.
```

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

# Get Started with the OpenFeature Java SDK and Spring Boot

## Introduction

This walk-through teaches you the basics of using OpenFeature in Java on top of a Spring Boot Web application.

You'll learn how to:

> - Integrate OpenFeature Java SDK

Reviewer 2: ```suggestion
- Integrate OpenFeature the Java SDK
```

Author: Will correct to `Integrate the OpenFeature Java SDK`

## Walk-through

### Step 1: Create a Spring Boot Web application

To get started, visit [Spring Initializer](https://start.spring.io/). This website allows you to download a pre-configured
Spring boot application. 

For our use case, add **Spring Web** from the dependencies section and set the following **project metadata** values,

- Group: com
- Artifact: demo

Select a Java version matching your installation and download the application using the **GENERATE** button.
Extract the archive and open it in your favorite editor.

### Step 2: Add dependencies

Based on the selected build system for Spring Boot application, add OpenFeature Java SDK and flagd provider dependencies
to existing dependency management configuration. 

Given below are dependencies for Maven and Gradle with the latest dependency versions at the time of writing this walk-through.

<Tabs groupId="dependency">
<TabItem value="maven" label="Maven">

```xml
<dependency>
    <groupId>dev.openfeature</groupId>
    <artifactId>sdk</artifactId>
    <version>1.0.0</version>
</dependency>
<dependency>
    <groupId>dev.openfeature.contrib.providers</groupId>
    <artifactId>flagd</artifactId>

>     <version>0.5.1</version>

Reviewer: ```suggestion
    <version>0.5.2</version>
```
Please bump this version now that it's available.

</dependency>
```

</TabItem>
<TabItem value="gradle" label="Gradle">

```groovy
implementation 'dev.openfeature:sdk:1.0.0'
implementation 'dev.openfeature.contrib.providers:flagd:0.5.1'
```
</TabItem>
</Tabs>

### Step 3: Add code

The `OpenFeatureAPI` class is the main access point to OpenFeature SDK. This class is designed to act as a singleton. Let's define
a simple bean configuration class called `OpenFeatureBeans`, that allows us to inject the `OpenFeatureAPI` singleton into desired Spring components.

Create the Java class `OpenFeatureBeans.java` inside the package `com.demo` and add the following code,
```java
package com.demo;

import dev.openfeature.sdk.OpenFeatureAPI;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

@Configuration
public class OpenFeatureBeans {

    @Bean
    public OpenFeatureAPI OpenFeatureAPI() {
        final OpenFeatureAPI openFeatureAPI = OpenFeatureAPI.getInstance();

        return openFeatureAPI;

>     }

Reviewer 2: I think perhaps injecting a client is an improvement over the API itself. The client is typically the thing `application authors` will interact with. The API is useful too, but it's more for one-time top level configuration.

Author: My view on `OpenFeatureClient` is that they are specific to usage (ex:- REST endpoint specific, service specific) and is customizable as needed. And `OpenFeatureAPI ` is the singleton with "base configurations" which allows users to generate clients. 

IMO injecting `OpenFeatureAPI ` is the right thing and then users can generate clients on the use case and preserve them for long term use. Let me know if you have different view :)

Author: There is also the possibility to expose `Client` singletons from `OpenFeatureBeans` (ex:- HelloOpenFeatureClient with some metadata like name and version), but that could increase the complexity of the guide

Reviewer 2: Both are valid. Client's can be short lived or long, and we can create one per request, even. They are lightweight and generally you can create many. My thinking was just that I would probably want a client for a particular part of the application (package), and the controller(s) could share one for multiple requests since theirs no reason not to here. Additionally, since the API is a singleton that's globally accessible, injecting it as a bean is a bit redundant.

This is just opinion though, I don't mind either way. In a demo application like this either is fine.

Author: Thanks for the view and I too agree with your view. But for now, let's keep it and change it later on if we see the need

}
```

Then we can add the REST endpoint definition of `/hello` endpoint, along with dependency injection for `OpenFeatureAPI` instance.

> Create the Java class `RestHello.java` inside the package `com.demo` and add the following code,

Reviewer: ```suggestion
Create the Java class `RestHello.java` inside the package `com.demo` and add the following code.
```

```java
package com.demo;

import dev.openfeature.sdk.Client;
import dev.openfeature.sdk.OpenFeatureAPI;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RestController;

@RestController
public class RestHello {
    private final OpenFeatureAPI openFeatureAPI;

    @Autowired
    public RestHello(OpenFeatureAPI OFApi) {
        this.openFeatureAPI = OFApi;
    }
    
    @GetMapping("/hello")
    public String getHello() {
        final Client client = openFeatureAPI.getClient();

        // Evaluate welcome-message feature flag
        if (client.getBooleanValue("welcome-message", false)) {

>             return "Hello, welcome to website!";

Author: Good suggestion . Including this with the next commit

Providers are an important concept in OpenFeature because they are responsible for the flag evaluation itself.
As we saw in the previous step, OpenFeature without a provider always returns the default value. If we want to actually perform feature
flagging, we'll need to register a provider.

Create a new file named `flags.json` and add the following JSON. Notice that there's a flag called `welcome-message` which matches the flag
key defined in [step 3](#step-3-add-code). The `welcome-message` flag has `on` and `off` variants that return `true` and `false` respectively.
The state property controls whether the feature flag is active or not. Finally, the defaultVariant property controls the variant that should be returned. 
In this case, the defaultVariant is `off`, therefore the value `false` would be returned.

```json
{
  "$schema": "https://raw.githubusercontent.com/open-feature/schemas/main/json/flagd-definitions.json",
  "flags": {
    "welcome-message": {
      "variants": {
        "on": true,
        "off": false
      },
      "state": "ENABLED",
      "defaultVariant": "off"
    }
  }
}
```
> NOTE: This configuration is specific for flagd and varies across providers.

> With the flagd configuration and image in place, start flagd service with the following docker command.

Reviewer: ```suggestion
With the flagd configuration in place, start flagd service with the following docker command.
```

```sh
docker run -p 8013:8013 -v $(pwd)/:/etc/flagd/ -it ghcr.io/open-feature/flagd:latest start --uri /etc/flagd/flags.json
```

> Finally, let's add the required code change to `OpenFeatureBeans.java` class,

Reviewer: ```suggestion
Finally, let's add the required code change to `OpenFeatureBeans.java` class.
```

```diff
package com.demo;

+ import dev.openfeature.contrib.providers.flagd.FlagdProvider;
import dev.openfeature.sdk.OpenFeatureAPI;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

@Configuration
public class OpenFeatureBeans {

    @Bean
    public OpenFeatureAPI OpenFeatureAPI() {
        final OpenFeatureAPI openFeatureAPI = OpenFeatureAPI.getInstance();
        
+        // Use flagd as the OpenFeature provider and use default configurations
+        openFeatureAPI.setProvider(new FlagdProvider());

        return openFeatureAPI;
    }
}
```

### Step 7: Rerun the application

> Let's rerun our Java Spring Boot application,

Reviewer: ```suggestion
Let's rerun our Java Spring Boot application.
```

<Tabs groupId="running">
<TabItem value="maven" label="Maven">

```shell
mvn clean install
java -jar target/demo-0.0.1-SNAPSHOT.jar
```
</TabItem>
<TabItem value="gradle" label="Gradle">

```shell
./gradlew clean build
java -jar build/libs/demo-0.0.1-SNAPSHOT.jar
```
</TabItem>
</Tabs>

Revisit the endpoint [http://localhost:8080/hello](http://localhost:8080/hello) and you will see the same value.

Now let's change the feature flag in our `flags.json`, making `defaultVariant` to `on`

```diff
{
  "$schema": "https://raw.githubusercontent.com/open-feature/schemas/main/json/flagd-definitions.json",
  "flags": {
    "welcome-message": {
      "variants": {
        "on": true,
        "off": false
      },
      "state": "ENABLED",
-      "defaultVariant": "off"
+      "defaultVariant": "on"
    }
  }
}
```

> Revisit the endpoint [http://localhost:8080/hello](http://localhost:8080/hello) and you will be greeted with `Hello, welcome to website!`

Reviewer 2: ```suggestion
Revisit the endpoint [http://localhost:8080/hello](http://localhost:8080/hello) and you will be greeted with `Hello, welcome to this OpenFeature-enabled website!`
```
